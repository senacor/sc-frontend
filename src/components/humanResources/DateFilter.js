import React, { Component } from 'react';
import * as actions from '../../actions';
import { getSubFilter } from '../../reducers/selector';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ListItem from '@material-ui/core/ListItem/ListItem';
import List from '@material-ui/core/List/List';
import TextField from '@material-ui/core/TextField/TextField';
import IconButton from '@material-ui/core/IconButton/IconButton';
import Icon from '@material-ui/core/Icon/Icon';
import moment from 'moment';
import { withStyles } from '@material-ui/core';

let styles = {
  error: {
    font: 'small',
    color: '#ff0000'
  }
};

export class DateFilter extends Component {
  constructor(props) {
    super(props);
    const filter = Object.assign(
      {},
      { searchString: '', values: '' },
      this.props.filter
    );

    this.state = {
      filter: filter,
      values: filter.values,
      error: false
    };
  }

  showError = () => {
    this.setState({ error: true });
  };

  isValidFilter = values => {
    return moment(values.From).isBefore(moment(values.To));
  };

  setFilter = values => {
    let searchString = Object.keys(values)
      .map(function(key) {
        return `${this.props.filterBy}${key}=${values[key]}`;
      }, this)
      .join('&');

    let filter = {
      searchString,
      values: values
    };
    let payload = {
      filterGroup: this.props.filterGroup,
      filterBy: this.props.filterBy,

      filter: filter
    };

    if (searchString === '') {
      this.props.deleteFilter(payload);
    } else if (!this.isValidFilter(values)) {
      return false;
    } else {
      this.props.addFilter(payload);
    }
    return true;
  };

  onChange = key => event => {
    let newValues = Object.assign({}, this.state.values, {
      [key]: event.target.value
    });
    if (newValues[key] === '') {
      delete newValues[key];
    }

    this.setState({ values: newValues, error: false });
  };

  execute = () => {
    if (this.setFilter(this.state.values)) {
      this.props.closeFilter();
    } else {
      this.showError();
    }
  };

  render() {
    const { classes } = this.props;
    let { values } = this.state;
    return (
      <List>
        <ListItem>
          <TextField
            id="startDate"
            label="von"
            type="date"
            defaultValue={values.From}
            InputLabelProps={{
              shrink: true
            }}
            onChange={this.onChange('From')}
          />
          <TextField
            id="endDate"
            label="bis"
            type="date"
            defaultValue={values.To}
            InputLabelProps={{
              shrink: true
            }}
            onChange={this.onChange('To')}
          />
          <IconButton
            id="forwardButton"
            aria-label="forward"
            onClick={this.execute}
          >
            <Icon>forward</Icon>
          </IconButton>
        </ListItem>
        <ListItem className={classes.error}>
          {this.state.error === true ? (
            <div>Falsche Eingabe: von nach bis</div>
          ) : null}
        </ListItem>
      </List>
    );
  }
}

DateFilter.propTypes = {
  filterGroup: PropTypes.string.isRequired,
  filterBy: PropTypes.string.isRequired
};

export const StyledComponent = withStyles(styles)(DateFilter);
export default connect(
  (state, props) => ({
    filter: getSubFilter(props.filterGroup, props.filterBy)(state)
  }),
  {
    addFilter: actions.addFilter,
    deleteFilter: actions.deleteFilter
  }
)(StyledComponent);
