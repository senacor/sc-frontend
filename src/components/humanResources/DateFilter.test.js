import React from 'react';
import { DateFilter } from './DateFilter';
import { createShallow } from '@material-ui/core/test-utils';
import FILTER_GROUPS from './filterGroups';
import HR_ELEMENTS from './hrElements';

describe('DateFilter Component', () => {
  let shallow = createShallow({ dive: false });

  it('should match snapshot', () => {
    const mockCallBack = jest.fn();

    let component = shallow(
      <DateFilter
        filterGroup={FILTER_GROUPS.HR}
        filterBy={HR_ELEMENTS.EMPLOYEE}
        filter={{
          searchString: 'deadlineFrom=2018-01-01&deadlineTo=2018-12-31',
          values: { From: '2018-01-01', To: '2018-12-31' }
        }}
        addFilter={mockCallBack}
        deleteFilter={mockCallBack}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('should able to change Date from', () => {
    const deleteFilter = jest.fn();
    const addFilter = jest.fn();

    let component = shallow(
      <DateFilter
        filterGroup={FILTER_GROUPS.HR}
        filterBy={HR_ELEMENTS.DEADLINE}
        filter={{
          searchString: 'deadlineFrom=2018-01-01&deadlineTo=2018-12-31',
          values: { From: '2018-01-01', To: '2018-12-31' }
        }}
        addFilter={addFilter}
        deleteFilter={deleteFilter}
      />
    );
    let event = { target: { value: '2018-01-13' } };
    component.find('#startDate').simulate('change', event, 'From');

    expect(component.find('#startDate').props().defaultValue).toEqual(
      '2018-01-13'
    );
    expect(component.find('#endDate').props().defaultValue).toEqual(
      '2018-12-31'
    );
  });

  it('should able to change Date to', () => {
    const deleteFilter = jest.fn();
    const addFilter = jest.fn();

    let component = shallow(
      <DateFilter
        filterGroup={FILTER_GROUPS.HR}
        filterBy={HR_ELEMENTS.DEADLINE}
        filter={{
          searchString: 'deadlineFrom=2018-01-01&deadlineTo=2018-12-31',
          values: { From: '2018-01-01', To: '2018-12-31' }
        }}
        addFilter={addFilter}
        deleteFilter={deleteFilter}
      />
    );
    let event = { target: { value: '2018-05-30' } };
    component.find('#endDate').simulate('change', event, 'to');

    expect(component.find('#startDate').props().defaultValue).toEqual(
      '2018-01-01'
    );
    expect(component.find('#endDate').props().defaultValue).toEqual(
      '2018-05-30'
    );
  });

  it('should able to change Date to and change Filter', () => {
    const deleteFilter = jest.fn();
    const addFilter = jest.fn();
    const closeFilter = jest.fn();

    let component = shallow(
      <DateFilter
        filterGroup={FILTER_GROUPS.HR}
        filterBy={HR_ELEMENTS.DEADLINE}
        filter={{
          searchString: 'deadline_from=2018-01-01&deadline_to=2018-12-31',
          values: { from: '2018-01-01', to: '2018-12-31' }
        }}
        addFilter={addFilter}
        deleteFilter={deleteFilter}
        closeFilter={closeFilter}
      />
    );

    let payload = {
      filterGroup: FILTER_GROUPS.HR,
      filterBy: HR_ELEMENTS.DEADLINE,
      filter: {
        searchString: `${HR_ELEMENTS.DEADLINE}From=2018-01-01&${
          HR_ELEMENTS.DEADLINE
        }To=2018-12-31`,
        values: { From: '2018-01-01', To: '2018-12-31' }
      }
    };

    component.find('#forwardButton').simulate('click');
    expect(addFilter).toHaveBeenCalledTimes(1);
    expect(addFilter).toHaveBeenCalledWith(payload);
  });
});
