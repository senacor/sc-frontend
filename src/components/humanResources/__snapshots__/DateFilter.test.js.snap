// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateFilter Component should match snapshot 1`] = `
<WithStyles(List)>
  <WithStyles(ListItem)>
    <TextField
      InputLabelProps={
        Object {
          "shrink": true,
        }
      }
      defaultValue="2018-01-01"
      id="startDate"
      label="von"
      onChange={[Function]}
      placeholder="JJJJ-MM-TT"
      required={false}
      select={false}
      type="date"
      variant="standard"
    />
    <TextField
      InputLabelProps={
        Object {
          "shrink": true,
        }
      }
      defaultValue="2018-12-31"
      id="endDate"
      label="bis"
      onChange={[Function]}
      placeholder="JJJJ-MM-TT"
      required={false}
      select={false}
      type="date"
      variant="standard"
    />
    <WithStyles(IconButton)
      aria-label="forward"
      id="forwardButton"
      onClick={[Function]}
    >
      <WithStyles(Icon)>
        forward
      </WithStyles(Icon)>
    </WithStyles(IconButton)>
  </WithStyles(ListItem)>
  <WithStyles(ListItem)
    className="DateFilter-error-1"
  />
</WithStyles(List)>
`;
